spock-core/src/main/java/org/spockframework/runtime/extension/IStore.java

```diff
@@ -0,0 +1,384 @@
+   * <p>
+   * See {@link #getOrComputeIfAbsent(Object, Function, Class)} for
+   * further details.
+   * <p>
+   * If {@code type} implements {@link AutoCloseable}
+   * the {@code close()} method will be invoked on the stored object when
+   * the store is closed.
+   *
+   * @param type the type of object to retrieve; never {@code null}
+   * @param <V>  the key and value type
+   * @return the object; never {@code null}
+   * @see #getOrComputeIfAbsent(Object, Function)
+   * @see #getOrComputeIfAbsent(Object, Function, Class)
+   * @see AutoCloseable
+   */
+  default <V> V getOrComputeIfAbsent(Class<V> type) {
+    return getOrComputeIfAbsent(type, ReflectionUtil::newInstance, type);
+  }
+
+  /**
+   * Get the value that is stored under the supplied {@code key}.
+   * <p>
+   * If no value is stored in the current {@link IStore}
+   * for the supplied {@code key}, ancestors of the context will be queried
+   * for a value with the same {@code key} in the {@code Namespace} used
+   * to create this store. If no value is found for the supplied {@code key},
+   * a new value will be computed by the {@code defaultCreator} (given
+   * the {@code key} as input), stored, and returned.
+   * <p>
+   * For greater type safety, consider using
+   * {@link #getOrComputeIfAbsent(Object, Function, Class)} instead.
+   * <p>
+   * If the created value is an instance of {@link AutoCloseable}
+   * the {@code close()} method will be invoked on the stored object when
+   * the store is closed.
+   * <p>
+   * <b>Warning:</b> Be aware that choosing {@code V} must be done wisely as it is an unsafe operation. If you for
+   * example select {@code String} for {@code V} and then try to get a {@code List} object from the returned
+   * optional, you will get a {@link ClassCastException} at runtime.
+   *
+   * @param key            the key; never {@code null}
+   * @param defaultCreator the function called with the supplied {@code key}
+   *                       to create a new value; never {@code null}
+   * @param <K>            the key type
+   * @param <V>            the value type
+   * @return the value; potentially {@code null}
+   * @see #getOrComputeIfAbsent(Class)
+   * @see #getOrComputeIfAbsent(Object, Function, Class)
+   * @see AutoCloseable
+   */
+
+  @Nullable
+  <K, V> V getOrComputeIfAbsent(K key, Function<K, V> defaultCreator);
+
+  /**
+   * Get the value of the specified required type that is stored under the
+   * supplied {@code key}.
+   * <p>
+   * If no value is stored in the current {@link IStore}
+   * for the supplied {@code key}, ancestors of the context will be queried
+   * for a value with the same {@code key} in the {@code Namespace} used
+   * to create this store. If no value is found for the supplied {@code key},
+   * a new value will be computed by the {@code defaultCreator} (given
+   * the {@code key} as input), stored, and returned.
+   * <p>
+   * If {@code requiredType} implements {@link AutoCloseable}
```

Comment:
`requiredType` is irrelevant, isn't it? Important is the actual type which could be a subtype which could be `AutoCloseable`.